File: apps/api/src/domain/webhooks.rs
//! Outbound webhook dispatch.
//!
//! `dispatch` finds the active webhooks in a project subscribed to an event and
//! enqueues one `deliver_webhook` job each. The jobs worker signs and POSTs the
//! payload, with the worker's built-in retry/backoff. Best-effort: callers
//! log-and-ignore so dispatch never blocks or fails the request path.

use serde_json::json;
use sqlx::PgPool;
use uuid::Uuid;

use crate::AppResult;

/// Event names a webhook can subscribe to (the `events` array on a webhook).
pub const EVENTS: &[&str] = &[
    "task.created",
    "task.updated",
    "task.moved",
    "task.deleted",
    "comment.created",
];

/// Enqueue a delivery job for each matching webhook. Returns how many.
///
/// The job carries the raw `(event, data)`; the worker formats the body per
/// the webhook's `target_type` at delivery time (ADR 0002). A row is
/// deliverable if it's a chat target (URL is the credential) or a generic
/// `outbound` target with a configured signing secret.
pub async fn dispatch(
    pool: &PgPool,
    project_id: Uuid,
    event: &str,
    data: serde_json::Value,
) -> AppResult<u64> {
    let ids: Vec<Uuid> = sqlx::query_scalar(
        r#"
        SELECT id FROM webhooks
        WHERE  project_id = $1 AND active = true AND deleted_at IS NULL
          AND  $2 = ANY(events)
          AND  (target_type <> 'outbound' OR secret_ciphertext IS NOT NULL)
        "#,
    )
    .bind(project_id)
    .bind(event)
    .fetch_all(pool)
    .await?;

    for id in &ids {
        enqueue(pool, *id, event, &data).await?;
    }
    Ok(ids.len() as u64)
}

/// Enqueue a single delivery job (shared by dispatch + send-test).
async fn enqueue(
    pool: &PgPool,
    webhook_id: Uuid,
    event: &str,
    data: &serde_json::Value,
) -> AppResult<()> {
    sqlx::query(r#"INSERT INTO jobs (id, kind, payload) VALUES ($1, 'deliver_webhook', $2)"#)
        .bind(Uuid::now_v7())
        .bind(json!({ "webhook_id": webhook_id, "event": event, "data": data }))
        .execute(pool)
        .await?;
    Ok(())
}

/// Enqueue a synthetic `test` delivery to one webhook, bypassing the event
/// subscription filter. Lets an admin confirm a target is wired up; the
/// result lands in `webhook_deliveries` like any real delivery.
pub async fn enqueue_test(pool: &PgPool, webhook_id: Uuid) -> AppResult<()> {
    let data = json!({ "key": "TEST-1", "message": "hello from Sprintly" });
    enqueue(pool, webhook_id, "test", &data).await
}

File: apps/web/lib/webhooks.ts
// Per-project outbound webhooks (F2). A webhook is a generic signed endpoint
// ("outbound") or a Slack/Discord target that gets a formatted message.

import { api } from "./api";

export type WebhookTarget = "outbound" | "slack" | "discord";

/// Events a webhook can subscribe to (mirrors domain::webhooks::EVENTS).
export const WEBHOOK_EVENTS = [
  "task.created",
  "task.updated",
  "task.moved",
  "task.deleted",
  "comment.created",
] as const;

export type Webhook = {
  id: string;
  project_id: string;
  url: string;
  target_type: WebhookTarget;
  events: string[];
  active: boolean;
  last_delivery_at: string | null;
  last_status: number | null;
  created_at: string;
};

export type WebhookDelivery = {
  id: string;
  event: string;
  status_code: number | null;
  ok: boolean;
  error: string | null;
  attempt: number;
  created_at: string;
};

export const listWebhooks = (key: string) =>
  api<{ items: Webhook[] }>(`/projects/${encodeURIComponent(key)}/webhooks`).then(
    (r) => r.items,
  );

export const createWebhook = (
  key: string,
  body: {
    url: string;
    target_type: WebhookTarget;
    events: string[];
    secret?: string;
  },
) =>
  api<Webhook>(`/projects/${encodeURIComponent(key)}/webhooks`, {
    method: "POST",
    body,
  });

export const updateWebhook = (
  id: string,
  body: { url?: string; events?: string[]; active?: boolean; secret?: string },
) => api<Webhook>(`/webhooks/${id}`, { method: "PATCH", body });

export const deleteWebhook = (id: string) =>
  api<void>(`/webhooks/${id}`, { method: "DELETE" });

export const listDeliveries = (id: string) =>
  api<{ items: WebhookDelivery[] }>(`/webhooks/${id}/deliveries`).then((r) => r.items);

export const sendTestWebhook = (id: string) =>
  api<void>(`/webhooks/${id}/test`, { method: "POST" });
